chronos/streams/tlsstream.nim

```diff
@@ -0,0 +1,334 @@
+        continue
+
+  except CancelledError:
+    rstream.state = AsyncStreamState.Stopped
+
+  finally:
+    # Perform TLS cleanup procedure
+    sslEngineClose(engine)
+    # Becase tlsWriteLoop() is ephemeral, but we still need to keep stream state
+    # consistent.
+    rstream.buffer.forget()
+    rstream.stream = nil
+
+proc newTlsClientAsyncStream*(rsource: AsyncStreamReader,
+                              wsource: AsyncStreamWriter,
+                              serverName: string = "",
+                              bufferSize = SSL_BUFSIZE_BIDI,
+                              minVersion = TLSVersion.TLS11,
+                              maxVersion = TLSVersion.TLS12,
+                              flags: set[TlsFlags] = {}): TlsAsyncStream =
+  ## Create new TLS asynchronous stream using reading stream ``rsource``,
+  ## writing stream ``wsource``.
+  ##
+  ## You can specify remote server name using ``serverName``, if while
+  ## handshake server reports different name you will get an error. If
+  ## ``serverName`` is empty string, remote server name checking will be
+  ## disabled.
+  ##
+  ## ``bufferSize`` - is SSL/TLS buffer which is used for encoding/decoding
+  ## incoming data.
+  ##
+  ## ``minVersion`` and ``maxVersion`` are TLS versions which will be used
+  ## for handshake with remote server. If server's version will be lower then
+  ## ``minVersion`` of bigger then ``maxVersion`` you will get an error.
+  ##
+  ## ``flags`` - custom TLS connection flags.
+  result = new TlsAsyncStream
+  var reader = new TlsStreamReader
+  reader.kind = TlsStreamKind.Client
+  var writer = new TlsStreamWriter
+  writer.kind = TlsStreamKind.Client
+  var switchToWriter = newAsyncEvent()
+  var switchToReader = newAsyncEvent()
+  reader.stream = result
+  writer.stream = result
+  reader.switchToReader = switchToReader
+  reader.switchToWriter = switchToWriter
+  writer.switchToReader = switchToReader
+  writer.switchToWriter = switchToWriter
+  result.reader = reader
+  result.writer = writer
+  reader.ccontext = addr result.context
+  writer.ccontext = addr result.context
+
+  if TLSFlags.NoVerifyHost in flags:
+    sslClientInitFull(addr result.context, addr result.x509, nil, 0)
+    initNoAnchor(addr result.xwc, addr result.x509.vtable)
+    sslEngineSetX509(addr result.context.eng, addr result.xwc.vtable)
+  else:
+    sslClientInitFull(addr result.context, addr result.x509,
+                      unsafeAddr MozillaTrustAnchors[0],
+                      len(MozillaTrustAnchors))
+
+  let size = max(SSL_BUFSIZE_BIDI, bufferSize)
+  result.sbuffer = newSeq[byte](size)
+  sslEngineSetBuffer(addr result.context.eng, addr result.sbuffer[0],
+                     uint(len(result.sbuffer)), 1)
+  sslEngineSetVersions(addr result.context.eng, uint16(minVersion),
+                       uint16(maxVersion))
+
+  if TLSFlags.NoVerifySN in flags:
+    let err = sslClientReset(addr result.context, "", 0)
```

Comment:
Are you sure the servername has to be empty here? What about the usecase when we want a self-signed cert on a multidomain+signgleip server?

Reply:
```c
/**
 * \brief Prepare or reset a client context for a new connection.
 *
 * The `server_name` parameter is used to fill the SNI extension; the
 * X.509 "minimal" engine will also match that name against the server
 * names included in the server's certificate. If the parameter is
 * `NULL` then no SNI extension will be sent, and the X.509 "minimal"
 * engine (if used for server certificate validation) will not check
 * presence of any specific name in the received certificate.
 *
 * Therefore, setting the `server_name` to `NULL` shall be reserved
 * to cases where alternate or additional methods are used to ascertain
 * that the right server public key is used (e.g. a "known key" model).
 *
 * If `resume_session` is non-zero and the context was previously used
 * then the session parameters may be reused (depending on whether the
 * server previously sent a non-empty session ID, and accepts the session
 * resumption). The session parameters for session resumption can also
 * be set explicitly with `br_ssl_engine_set_session_parameters()`.
 *
 * On failure, the context is marked as failed, and this function
 * returns 0. A possible failure condition is when no initial entropy
 * was injected, and none could be obtained from the OS (either OS
 * randomness gathering is not supported, or it failed).
 *
 * \param cc               client context.
 * \param server_name      target server name, or `NULL`.
 * \param resume_session   non-zero to try session resumption.
 * \return  0 on failure, 1 on success.
 */
int br_ssl_client_reset(br_ssl_client_context *cc,
	const char *server_name, int resume_session);
```

Reply:
NoVerifySN (Server Name Identification is OFF) is not same as NoVerifyHost (Server Certificate Verification is OFF)

Reply:
Oh I missed that. But in such case the NoVerifySN flag seems to be redundant to `serverName` argument, which should be just set to `""`. Why not just remove the flag altogether. Besides its name is pretty cryptic.

Reply:
Setting `serverName` to empty string is very visible, if you make the argument non-optional. `NoVerifySN` just doesn't tell anything (SN is a Serial Number to me), it only adds more useless code to review and to execute. Here's a nice API example:
```nim
proc greet(name: string, nameIsCF: bool = true) =
  ## Greets name.
  ## If `nameIsCF` is true, the `name` argument is ignored,
  ## and cheatfate is greeted. We've made sure to carefully precalculate
  ## the most recommended default value for this argument.
  if nameIsCheatfate:
    echo "Hello Cheatfate"
  else:
    echo "Hello ", name
```